common4j/src/main/com/microsoft/identity/common/java/cache/MsalOAuth2TokenCache.java

```diff
@@ -367,10 +366,11 @@ public ICacheRecord save(@NonNull final GenericOAuth2Strategy oAuth2Strategy,
 
         // Save the Account and Credentials...
         saveAccounts(accountToSave);
-        saveCredentialsInternal(accessTokenToSave, refreshTokenToSave, idTokenToSave);
-
-        // Remove old refresh tokens (except for the one we just saved) if it's MRRT or FRT
-        removeAllRefreshTokensExcept(accountToSave, refreshTokenToSave);
+        synchronized(this) {
+            saveCredentialsInternal(accessTokenToSave, refreshTokenToSave, idTokenToSave);
+            // Remove old refresh tokens (except for the one we just saved) if it's MRRT or FRT
+            removeAllRefreshTokensExcept(accountToSave, refreshTokenToSave);
```

Comment:
I'm wondering if this is enough. 

It would protect this specific method from race-condition itself, but it won't protect any other potential cache-related race condition issue. 

What if saveCredentialsInternal is invoked by other methods when this synchronize block is operating? (i.e setSingleSignOnState is also calling both functions).

Maybe we need some sort of read-write lock for the class' public interfaces?

Reply:
That probably is a bigger work item and out of scope for current fix, 
Shared Preference does provide some support for that, but we might need to think from our usage perspective, especially for cases where we expect two separate operations (add and remove) to happen atomically